---
title: "R Notebook"
output:
  html_document: default
  html_notebook: default
---

## Normal liver gene expression profile

- GTEx median RPKM by tissue (V6p): 119 donors
- PCAWG FPKM UQ: 69 donors

```{r, eval=FALSE}
library(data.table)
source('../figures/ggplot_theme.R')
gtex = read.table('~/Downloads/GTEx_Analysis_v6p_RNA-seq_RNA-SeQCv1.1.8_gene_median_rpkm.gct.gz',
                  header=TRUE, stringsAsFactors = FALSE, skip = 2, sep='\t')
rnameta = read.table('~/DriverPower/figures/ALB/rnaseq_metadata.lite.tsv', header=T, sep='\t', stringsAsFactors = F)
rnameta = rnameta[rnameta$project_code %in% c('LIHC-US', 'LIRI-JP'),]
table(rnameta$is_tumour)
pcawg = fread('~/analysis/ALB/tophat_star_fpkm_uq.v2_aliquot_gl.tsv', select = c('feature', rnameta$aliquot_id[rnameta$is_tumour=='no']))
sum(pcawg$feature %in% gtex$Name)
sum(gtex$Name %in% pcawg$feature)
# calculate median FPKM-UQ
pcawg = as.data.frame(pcawg)
pcawg['median'] = apply(pcawg[,-1], 1, median)
liver.exp = merge(pcawg[,c('feature', 'median')], gtex[,c('Name', 'Description', 'Liver')], by.x='feature', by.y='Name', all.x = TRUE)
liver.exp = liver.exp[,c('feature', 'Description', 'median', 'Liver')]
colnames(liver.exp) = c('gene_id', 'gene_symbol', 'pcawg', 'gtex')
# add pval and qval for DE
rnameta = read.table('../figures/ALB/rnaseq_metadata.lite.tsv', header=T, sep='\t', stringsAsFactors = F)
donor_id = read.table('../figures/ALB/Liver-HCC.tumor_barcode.txt', header=F, sep=' ', stringsAsFactors = F)
rnameta = rnameta[rnameta$icgc_donor_id %in% donor_id$V2,]
pcawg = fread('~/Downloads/tophat_star_fpkm_uq.v2_aliquot_gl.tsv', select = c('feature', rnameta$aliquot_id))
pcawg = as.data.frame(pcawg)
pcawg.tumor = pcawg[, c('feature', rnameta$aliquot_id[rnameta$is_tumour=='yes'])]
pcawg.tumor['pcawg.tumor'] = apply(pcawg.tumor[,-1], 1, median)
liver.exp = merge(liver.exp, pcawg.tumor[,c('feature', 'pcawg.tumor')], by.x='gene_id', by.y='feature')
liver.exp['log2FC'] = log2(liver.exp$pcawg.tumor) - log2(liver.exp$pcawg)
liver.exp['p'] = as.numeric(NA)
tumors = rnameta$aliquot_id[rnameta$is_tumour=='yes']
normals = rnameta$aliquot_id[rnameta$is_tumour=='no']
for (idx in 1:nrow(pcawg)){
  g = pcawg$feature[idx]
  pval = wilcox.test(as.numeric(pcawg[idx, tumors]), as.numeric(pcawg[idx, normals]))$p.value
  liver.exp[liver.exp$gene_id==g, 'p'] = pval
}
liver.exp['q'] = p.adjust(liver.exp$p, 'BH')
write.table(liver.exp, './liver.expression.tsv', sep='\t', row.names = F, quote = F)
```

## Top 1/1000 genes

```{r}
library(ggplot2)
source('../figures/ggplot_theme.R')
liver.exp = read.table('./liver.expression.tsv', sep = '\t', header = TRUE, stringsAsFactors = FALSE)
ggplot(liver.exp, aes(x=log10(pcawg+1), y=log10(gtex+1))) + geom_point() + theme_Publication()
# 1/1000 gene
pcawg_cutoff = quantile(liver.exp$pcawg, 0.999)
gtex_cutoff = quantile(liver.exp$gtex, 0.999, na.rm = TRUE)
pcawg_genes = subset(liver.exp, pcawg>=pcawg_cutoff)
gtex_genes = subset(liver.exp, gtex>=gtex_cutoff)
pcawg_genes = pcawg_genes[order(pcawg_genes$pcawg, decreasing = T),]
gtex_genes = gtex_genes[order(gtex_genes$gtex, decreasing = T),]
both_genes = intersect(gtex_genes$gene_symbol, pcawg_genes$gene_symbol) # 53 genes
# remove mito. genes
is_mt = substr(both_genes, 1, 2) == 'MT'
mt_genes = both_genes[is_mt]
mt_genes = mt_genes[-17] # 16 genes
nuc_genes = both_genes[!both_genes %in% mt_genes] # 27 genes
cat(nuc_genes)
```

```{r, fig.height=6, fig.width=6}
library(VennDiagram)
venn.plot = draw.pairwise.venn(nrow(pcawg_genes), nrow(gtex_genes), length(both_genes), category = c('PCAWG', 'GTEx'))
grid.draw(venn.plot)
```

## Expression of `nuc_genes` across normal tissues
```{r, fig.height=6, fig.width=8}
gtex = read.table('~/Downloads/GTEx_Analysis_v6p_RNA-seq_RNA-SeQCv1.1.8_gene_median_rpkm.gct.gz',
                  header=TRUE, stringsAsFactors = FALSE, skip = 2, sep='\t', check.names = F)
# pdf('./GTEx.tissue.specific.expression.pdf', height = 7, width = 9)
for (g in nuc_genes){
  dat = as.data.frame(t(gtex[gtex$Description==g, 3:ncol(gtex)]))
  dat[,'tissue'] = row.names(dat)
  colnames(dat)[1] = 'MedianRPKM'
  dat[,'is_liver'] = ifelse(dat$tissue=='Liver', 'Liver', 'Others')

  p = ggplot(dat, aes_string(x='tissue', y='MedianRPKM', fill='is_liver')) + geom_bar(stat = 'identity') +
    theme_Publication() + scale_fill_Publication1() + ylab('Median RPKM') +
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0),
          axis.title.x = element_blank(), legend.title = element_blank()) +
    annotate('text', x=10, y=max(dat$MedianRPKM)*0.8, label=g, size=10)
  print(p)
}

g = 'APOB'
# dev.off()
```

## Expression of `nuc_genes` between tumor and normal
```{r}
library(data.table)
rnameta = read.table('../figures/ALB/rnaseq_metadata.lite.tsv', header=T, sep='\t', stringsAsFactors = F)
donor_id = read.table('../figures/ALB/Liver-HCC.tumor_barcode.txt', header=F, sep=' ', stringsAsFactors = F)
rnameta = rnameta[rnameta$icgc_donor_id %in% donor_id$V2,]
pcawg = fread('~/Downloads/tophat_star_fpkm_uq.v2_aliquot_gl.tsv', select = c('feature', rnameta$aliquot_id))
pcawg = as.data.frame(pcawg)

# pdf('./PCAWG.Liver-HCC.expression.normal.vs.tumor.pdf')
for (g in nuc_genes) {
  g.id = pcawg_genes$gene_id[pcawg_genes$gene_symbol==g]
  g.exp = as.data.frame(t(pcawg[pcawg$feature==g.id, -1]))
  g.exp['aliquot_id'] = row.names(g.exp)
  colnames(g.exp)[1] = 'FPKM.UQ'
  g.exp = merge(g.exp, rnameta, by='aliquot_id')
  pval = wilcox.test(FPKM.UQ ~ is_tumour, data = g.exp)$p.value
  p = ggplot(g.exp, aes(x=is_tumour, y=FPKM.UQ)) + geom_boxplot() +
    ggtitle(paste(g, '\np.value = ', signif(pval,5), sep='')) +
    theme_Publication()
  print(p)
}
# dev.off()

# ACTB
g = 'ACTB'

```